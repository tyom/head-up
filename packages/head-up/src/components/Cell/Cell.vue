<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{_toggled: showSettings}"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h"/>
      </button>
    </header>
    <div class="container">
      <div :class="bodyClass" class="body">
        <slot v-if="$slots.default"/>
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings"/>
      </transition>
    </div>
  </div>
</template>

<script>
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && this.isEditing();
    },
    bodyClass() {
      const content = this.$slots.default ? this.$slots.default : this.content;
      return {
        [`u-grid-${content.length}-x`]: !!content.length,
        padded: this.padded,
      };
    },
  },
  inject: ['isEditing'],
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  border: 1px solid #0005;
  transition: 0.3s;
  display: flex;
  flex-direction: column;
  user-select: none;

  &:hover {
    transition: 0.1s;
  }
}

.header {
  font-size: 0.55em;
  color: #fff6;
  position: relative;
  text-transform: uppercase;
  background-color: #0003;
  padding: 0.7em 1em;
  line-height: 1;
  justify-content: center;
  display: flex;
  align-items: center;
}

.header-button {
  border: 0;
  padding: 0.5em 1em;
  background: none;
  color: #fff;
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  bottom: 0;
  opacity: 0.7;
  outline: 0;

  &:hover {
    opacity: 1;
  }

  &:focus-visible {
    background-color: rgba(#20416d, 0.7);
    box-shadow: 0 0 0 1px #fff;
  }

  &._toggled {
    opacity: 1;
    background-color: #20416d;
  }

  &:active {
    opacity: 1;
    transition: none;
    background-color: #20416d;
  }
}

.title {
  font-size: inherit;
}

.container {
  position: relative;
  flex: 1;
  display: flex;
}

.body {
  grid-gap: 1px;
  background-color: rgba(#000, 0.2);
  flex: 1;
  justify-content: stretch;
  align-items: stretch;
}

.body > * {
  background-color: rgba(#fff, 0.03);
}

.padded > * {
  padding: 0.5em;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  transition: 0.3s;
}
.revealSettings-enter,
.revealSettings-leave-to {
  opacity: 0;
}
</style>
