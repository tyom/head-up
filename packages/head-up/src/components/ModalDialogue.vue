<template>
  <transition name="reveal">
    <div v-if="show" class="ModalDialogue">
      <div class="backdrop" @click="$emit('close')"/>
      <div class="container">
        <header v-if="heading">
          <h2 class="title">{{ heading }}</h2>
        </header>
        <slot/>
      </div>
    </div>
  </transition>
</template>

<script>
export default {
  props: {
    show: {
      type: Boolean,
      default: false,
    },
    heading: {
      type: String,
      default: '',
    },
  },
};
</script>

<style scoped>
.ModalDialogue {
  position: fixed;
  z-index: 2;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.backdrop {
  background-color: #000b;
  height: 100%;
  width: 100%;
}

.container {
  position: absolute;
  border-radius: 0.3em;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  max-width: 60vw;
  max-height: 60vh;
  margin: auto;
  z-index: 1;
  background-color: #fff;
  color: #444;
  padding: 1em;
}

header {
  padding-bottom: 1em;
  margin-bottom: 1em;
  border-bottom: 0.1em solid #ccc;
}

.reveal-enter-active,
.reveal-leave-active {
  transition: 1s;
}
.reveal-enter-active {
  & .backdrop {
    transition: 0.5s;
  }

  & .container {
    transition: 0.2s 0.1s;
  }
}
.reveal-leave-active {
  & .backdrop {
    transition: 0.3s 0.1s;
  }

  & .container {
    transition: 0.1s;
  }
}

.reveal-enter,
.reveal-leave-to {
  & .backdrop {
    opacity: 0;
  }

  & .container {
    opacity: 0;
    transform: translateY(-50%);
  }
}
</style>
