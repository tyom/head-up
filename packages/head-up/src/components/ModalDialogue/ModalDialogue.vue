<template>
  <transition name="reveal">
    <div
      class="ModalDialogue"
      v-shortkey="['esc']"
      @shortkey="handleClose"
    >
      <div class="backdrop" @click="handleClose"/>
      <div class="container">
        <header v-if="heading">
          <h2 class="title">{{ heading }}</h2>
        </header>
        <slot/>
      </div>
    </div>
  </transition>
</template>

<script>
export default {
  props: {
    heading: {
      type: String,
      default: '',
    },
  },
  methods: {
    handleClose() {
      this.$emit('close');
    },
  },
};
</script>

<style scoped>
.ModalDialogue {
  position: fixed;
  z-index: 2;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.backdrop {
  background-color: #000b;
  height: 100%;
  width: 100%;
}

.container {
  position: absolute;
  border-radius: 0.3em;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  max-width: 60vw;
  max-height: 60vh;
  margin: auto;
  z-index: 1;
  background-color: #222;
  border: 2px solid #fff4;
  color: #ccc;
  padding: 1em;
}

header {
  padding: 1em;
  margin: -1em -1em 1em;
  background-color: #0002;
  border-bottom: 0.1em solid #fff2;
}

.reveal-enter-active,
.reveal-leave-active {
  transition: 1s;
}
.reveal-enter-active {
  & .backdrop {
    transition: 0.5s;
  }

  & .container {
    transition: 0.2s 0.1s;
  }
}
.reveal-leave-active {
  & .backdrop {
    transition: 0.3s 0.1s;
  }

  & .container {
    transition: 0.1s;
  }
}

.reveal-enter,
.reveal-leave-to {
  & .backdrop {
    opacity: 0;
  }

  & .container {
    opacity: 0;
    transform: translateY(-50%);
  }
}
</style>
