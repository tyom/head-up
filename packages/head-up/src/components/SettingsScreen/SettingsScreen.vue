<template>
  <form class="settings">
    <ul>
      <li v-for="(setting, key) in settings" :key="key">
        <VSwitchToggle
          v-model="settingsState[key].value"
          :label="setting.label"
          :name="key"
        />
      </li>
    </ul>
  </form>
</template>

<script>
export default {
  props: {
    settings: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      settingsState: this.settings,
    };
  },
  watch: {
    settingsState: {
      deep: true,
      handler() {
        this.$emit('save', this.settingsState);
      },
    },
  },
};
</script>

<style scoped>
ul {
  list-style: none;
  padding: 0;
  margin: 0;
}

li + li {
  margin-top: 1em;
}
</style>
