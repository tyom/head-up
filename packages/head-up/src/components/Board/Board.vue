<template>
  <div :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && isEditing()"
        :board="$props"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div
          v-for="(cell, idx) in cells"
          :key="idx"
          class="cell-placeholder"
        />
      </template>
      <slot v-else-if="!cells.length" />
      <template v-else>
        <Cell
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      default: '',
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    layoutClass() {
      const cells = this.$slots.default ? this.$slots.default : this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  inject: ['isEditing'],
};
</script>

<style scoped>
.Board {
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
