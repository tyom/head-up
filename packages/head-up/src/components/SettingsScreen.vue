<template>
  <ModalDialogue
    :show="show"
    :heading="heading"
    @close="$emit('close')"
  >
    <form class="settings">
      <p>
        <VSwitchToggle v-model="smoothScrollingValue" label="Smooth scrolling"/>
      </p>
    </form>
  </ModalDialogue>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import ModalDialogue from './ModalDialogue';

export default {
  components: {
    ModalDialogue,
  },
  props: {
    show: {
      type: Boolean,
      default: false,
    },
    heading: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      smoothScrollingValue: this.smoothScrolling,
    };
  },
  computed: {
    ...mapState(['smoothScrolling']),
  },
  watch: {
    // TODO This appears to be some async issue as the value should be
    // initialised in data function
    smoothScrolling: {
      immediate: true,
      handler(newVal) {
        this.smoothScrollingValue = newVal;
      },
    },
    smoothScrollingValue(newVal) {
      this.UPDATE_SETTING({
        smoothScrolling: newVal,
      });
    },
  },
  methods: {
    ...mapActions(['UPDATE_SETTING']),
  },
};
</script>
