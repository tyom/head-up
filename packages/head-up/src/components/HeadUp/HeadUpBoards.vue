<template>
  <div class="HeadUpBoards">
    <Board
      v-for="(board, idx) in boards"
      ref="activeBoardEl"
      v-bind="board"
      :editable="!board.isReadOnly"
      :key="board.title + board.id + idx"
    />
    <slot/>
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
  },
};
</script>

<style scoped>
.HeadUpBoards {
  flex: 1;
  overflow: hidden;
}
</style>
