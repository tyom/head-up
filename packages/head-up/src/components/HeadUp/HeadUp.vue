<template>
  <div
    v-shortkey="shortkeys"
    :class="rootClass"
    class="HeadUp"
    @shortkey="handleKeys"
  >
    <Sidebar
      v-if="!hideSidebar"
      :visible="state.showSidebar"
      @toggle="handleSidebarToggle"
      @board:add="handleAddBoard"
      @board:remove="handleRemoveBoard"
      @board:activate="handleActivateBoard"
      @board:edit="handleToggleEdit"
      @modal:settings="toggleSettingsScreen"
      @modal:help="toggleHelpScreen"
    />
    <div class="main">
      <HeadUpBoards :boards="state.boards" ref="boardsContainer">
        <slot/>
      </HeadUpBoards>
    </div>
    <ModalDialogue
      v-if="modal"
      :heading="modal.heading"
      @close="modal = null"
    >
      <HelpScreen
        v-if="modal.name === 'help'"
      />
      <SettingsScreen
        v-if="modal.name === 'settings'"
        :settings="state.settings"
        @save="handleSettingsSave"
      />
    </ModalDialogue>
  </div>
</template>

<script>
import { uniqueId, get } from 'lodash';
import { serializeSlot } from '../../transformers';
import ModalDialogue from '../ModalDialogue';
import Sidebar from '../Sidebar';
import SettingsScreen from '../SettingsScreen';
import HelpScreen from '../HelpScreen';
import HeadUpBoards from './HeadUpBoards';

export default {
  name: 'HeadUp',
  components: {
    HeadUpBoards,
    Sidebar,
    ModalDialogue,
    SettingsScreen,
    HelpScreen,
  },
  data() {
    return {
      state: {
        activeBoardIdx: 0,
        editMode: false,
        showSidebar: true,
        boards: this.boards,
        settings: {
          smoothScrolling: {
            type: 'toggle',
            label: 'Smooth scrolling',
            value: true,
          },
          persistState: {
            type: 'toggle',
            label: 'Persist to local storage',
            value: true,
          },
        },
      },
      persistedState: null,
      modal: null,
      shortkeys: {
        up: ['k'],
        down: ['j'],
        sidebar: ['s'],
        edit: ['e'],
        add: ['a'],
        settings: ['shift', '<'],
        help: ['shift', '?'],
      },
    };
  },
  props: {
    hideSidebar: {
      type: Boolean,
      default: false,
    },
    boards: {
      type: Array,
      default: () => [],
    },
  },
  computed: {
    rootClass() {
      return {
        _edit: this.state.editMode,
      };
    },
    boardSummary() {
      return [...this.state.boards, ...serializeSlot(this.$slots.default)];
    },
  },
  watch: {
    'state.activeBoardIdx'() {
      this.scrollToActiveBoard();
    },
    state: {
      deep: true,
      handler(newVal) {
        if (newVal.settings.persistState.value) {
          localStorage.setItem('headUp', JSON.stringify(newVal));
          return;
        }
        // save persistence setting but not the state
        const newState = JSON.parse(this.persistedState);
        newState.settings.persistState.value =
          newVal.settings.persistState.value;
        localStorage.setItem('headUp', JSON.stringify(newState));
      },
    },
  },
  provide() {
    return {
      isEditing: () => this.state.editMode,
      getBoardSummary: () => this.boardSummary,
      getActiveBoardIdx: () => this.state.activeBoardIdx,
      handleEditDone: this.handleEditDone,
      handleEditSave: this.handleEditSave,
    };
  },
  created() {
    this.persistedState = localStorage.getItem('headUp');
    if (this.persistedState) {
      this.state = JSON.parse(this.persistedState);
    }
  },
  mounted() {
    window.addEventListener('resize', () => this.scrollToActiveBoard(false));
    this.scrollToActiveBoard(false);
  },
  methods: {
    handleSidebarToggle(value) {
      this.state.showSidebar =
        typeof value === 'undefined' ? !this.state.showSidebar : value;
    },
    handleActivateBoard(value) {
      this.state.activeBoardIdx = value;
    },
    handleToggleEdit(value) {
      this.state.editMode =
        typeof value === 'undefined' ? !this.state.editMode : value;
    },
    handleEditSave(board) {
      this.state.boards = this.state.boards.map(
        x => (x.id === board.id ? board : x),
      );
    },
    handleEditDone(board) {
      this.handleEditSave(board);
      this.state.editMode = false;
    },
    handleAddBoard() {
      const newBoardTemplate = {
        title: `Board #${this.boardSummary.length + 1}`,
        id: uniqueId(),
        editable: true,
        cells: [
          {
            title: 'Cell #1',
          },
        ],
      };

      this.state.boards = [newBoardTemplate, ...this.state.boards];
      this.state.activeBoardIdx = 0;
      this.handleToggleEdit(true);
    },
    handleRemoveBoard(id) {
      const deletedIndex = this.state.boards.findIndex(x => x.id === id);
      this.state.boards = this.state.boards.filter(x => x.id !== id);

      if (this.state.activeBoardIdx === this.boardSummary.length) {
        this.state.activeBoardIdx--;
      } else if (this.state.activeBoardIdx === deletedIndex) {
        this.state.activeBoardIdx = deletedIndex;
      } else {
        this.state.activeBoardIdx = deletedIndex > 0 ? deletedIndex - 1 : 0;
      }
    },
    handleKeys(evt) {
      switch (evt.srcKey) {
        case 'up':
          this.activatePreviousBoard();
          break;
        case 'down':
          this.activateNextBoard();
          break;
        case 'sidebar':
          this.handleSidebarToggle();
          break;
        case 'edit':
          this.handleToggleEdit();
          break;
        case 'add':
          this.handleAddBoard();
          break;
        case 'settings':
          this.toggleSettingsScreen();
          break;
        case 'help':
          this.toggleHelpScreen();
          break;
        default:
      }
    },
    scrollToActiveBoard(useGlobalSetting = true) {
      if (!this.boardSummary.length) {
        return;
      }
      const boardEls = get(this.$refs, 'boardsContainer.$el.childNodes', []);
      const currentBoardEl = boardEls[this.state.activeBoardIdx];
      if (!currentBoardEl) {
        return;
      }
      const scrollOptions =
        useGlobalSetting && this.state.settings.smoothScrolling.value
          ? { behavior: 'smooth' }
          : {};
      currentBoardEl.scrollIntoView(scrollOptions);
    },
    activateNextBoard() {
      if (this.state.activeBoardIdx === this.boardSummary.length - 1) {
        this.handleActivateBoard(0);
        return;
      }
      this.handleActivateBoard(this.state.activeBoardIdx + 1);
    },
    activatePreviousBoard() {
      if (this.state.activeBoardIdx === 0) {
        this.handleActivateBoard(this.boardSummary.length - 1);
        return;
      }
      this.handleActivateBoard(this.state.activeBoardIdx - 1);
    },
    toggleHelpScreen() {
      this.toggleModal({ name: 'help', heading: 'Keyboard help' });
    },
    toggleSettingsScreen() {
      this.toggleModal({ name: 'settings', heading: 'Settings' });
    },
    toggleModal(modalData) {
      const existingModalName = this.modal && this.modal.name;
      this.modal = existingModalName === modalData.name ? null : modalData;
    },
    handleSettingsSave(payload) {
      this.state.settings = payload;
    },
  },
};
</script>

<style src="../../css/util.css"/>

<style scoped>
>>> * {
  &,
  &::before,
  &::after {
    box-sizing: border-box;
    margin: 0;
  }
}

.HeadUp {
  font-size: responsive 8px 23px;
  transition: 0.4s background-color;
  background-color: #222;
  color: #bbb;
  height: 100vh;
  width: 100vw;
  display: flex;
  align-items: center;

  &._edit {
    background-color: #0d1a2b;
  }
}

.main {
  height: 100vh;
  width: 100%;
  display: flex;
  flex-direction: column;
}
</style>
