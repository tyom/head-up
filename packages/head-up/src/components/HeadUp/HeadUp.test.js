import { merge } from 'lodash';
import ally from 'ally.js';
import ShortKey from 'vue-shortkey';
import Chance from 'chance';
import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import Board from '../Board';
import Cell from '../Cell';
import Sidebar from '../Sidebar';
import ModalDialogue from '../ModalDialogue';
import SettingsScreen from '../SettingsScreen';
import VSwitchToggle from '../VSwitchToggle';
import HeadUp from './HeadUp';
import HeadUpBoards from './HeadUpBoards';

jest.mock('ally.js');

const localVue = createLocalVue();
localVue.use(ShortKey);
localVue.component('VSwitchToggle', VSwitchToggle);

const scrollIntoViewSpy = jest.fn();

function mountHeadUp(options = {}) {
  return shallowMount(HeadUp, {
    localVue,
    ...options,
  });
}

function mountWithSlot(options = {}) {
  return mountHeadUp(
    merge(
      {
        localVue,
        slots: {
          default: '<Board id="b1"><Cell title="Cell #1"/></Board>',
        },
        stubs: {
          Cell,
          Board,
        },
      },
      options,
    ),
  );
}

function mountWithProps(options = {}) {
  return mountHeadUp(
    merge(
      {
        localVue,
        propsData: {
          boards: [
            {
              id: '1',
              cells: [{ title: 'Cell #1' }, { title: 'Cell #2' }],
            },
            {
              id: '2',
              cells: [{ title: 'Cell #3' }],
            },
            {
              id: '3',
              cells: [{ title: 'Cell #4' }],
            },
          ],
        },
        stubs: {
          Cell,
          Board,
          HeadUpBoards,
        },
      },
      options,
    ),
  );
}

beforeEach(() => {
  jest.resetAllMocks();
  Element.prototype.scrollIntoView = scrollIntoViewSpy;
  Chance.prototype.sentence = jest.fn(() => 'New board.');
});

test('render default', () => {
  const wrapper = mountHeadUp();
  expect(wrapper).toMatchSnapshot();
});

test('render without sidebar', () => {
  const wrapper = mountHeadUp({
    propsData: {
      hideSidebar: true,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render in edit mode', () => {
  const wrapper = mountHeadUp();

  wrapper.setData({
    state: {
      editMode: true,
    },
  });

  expect(wrapper).toMatchSnapshot();
});

test('render with slot', () => {
  const wrapper = mountWithSlot();

  expect(wrapper).toMatchSnapshot();
});

test('render with props', () => {
  const wrapper = mountWithProps();

  expect(wrapper).toMatchSnapshot();
});

test('scroll to activeBoard', () => {
  const wrapper = mountWithProps();

  wrapper.setData({
    state: {
      activeBoardId: '2',
    },
  });

  expect(scrollIntoViewSpy).toHaveBeenCalled();
});

test('add board in sidebar', () => {
  const wrapper = mountWithProps();
  const sidebar = wrapper.find(Sidebar);

  expect(wrapper.vm.state.boards.length).toEqual(3);
  sidebar.vm.$emit('board:add');

  expect(wrapper.vm.state.boards.length).toEqual(4);
  expect(wrapper.vm.state.boards[0].title).toEqual('new board');
  expect(wrapper.vm.state.boards[0].id).toEqual('new-board');
});

describe('remove board in sidebar', () => {
  test('remove active board', () => {
    const wrapper = mountWithProps();
    const sidebar = wrapper.find(Sidebar);

    expect(wrapper.vm.state.boards.length).toEqual(3);
    sidebar.vm.$emit('board:remove', '1');

    expect(wrapper.vm.state.boards.length).toEqual(2);

    process.nextTick(() => {
      expect(wrapper.vm.state.activeBoardId).toEqual('2');
      expect(wrapper.vm.activeBoardIndex).toEqual(0);
    });
  });

  test('remove last board in the list', () => {
    const wrapper = mountWithProps();
    const sidebar = wrapper.find(Sidebar);

    wrapper.setData({
      state: {
        activeBoardId: '3',
      },
    });
    expect(wrapper.vm.state.boards.length).toEqual(3);
    sidebar.vm.$emit('board:remove', '3');

    expect(wrapper.vm.state.boards.length).toEqual(2);
    process.nextTick(() => {
      expect(wrapper.vm.state.activeBoardId).toEqual('2');
      expect(wrapper.vm.activeBoardIndex).toEqual(1);
    });
  });

  test('remove board preceding the active one', () => {
    const wrapper = mountWithProps();
    const sidebar = wrapper.find(Sidebar);

    wrapper.setData({
      state: {
        activeBoardId: '2',
      },
    });
    expect(wrapper.vm.state.boards.length).toEqual(3);
    expect(wrapper.vm.activeBoardIndex).toEqual(1);
    sidebar.vm.$emit('board:remove', '1');

    expect(wrapper.vm.state.boards.length).toEqual(2);
    expect(wrapper.vm.activeBoardIndex).toEqual(0);
  });
});

describe('respond to board edits', () => {
  let wrapper;

  const newBoardData = {
    id: '1',
    title: 'Updated board',
  };

  beforeEach(() => {
    wrapper = mountWithProps();
  });

  test('save edit', () => {
    wrapper.vm.handleEditSave(newBoardData);
    expect(wrapper.vm.state.boards.find(x => x.id === '1')).toEqual(
      newBoardData,
    );
  });

  test('finish edit', () => {
    wrapper.vm.handleEditDone(newBoardData);
    expect(wrapper.vm.state.boards.find(x => x.id === '1')).toEqual(
      newBoardData,
    );
    expect(wrapper.vm.state.editMode).toBe(false);
  });
});

describe('modal dialogues', () => {
  let wrapper;
  let sidebar;

  beforeEach(() => {
    wrapper = mountWithSlot();
    sidebar = wrapper.find(Sidebar);
  });

  test('help', () => {
    expect(wrapper.vm.modal).toBeNull();
    sidebar.vm.$emit('modal:help');
    expect(wrapper.vm.modal.name).toEqual('help');
    expect(wrapper).toMatchSnapshot();

    sidebar.vm.$emit('modal:help');
    expect(wrapper.vm.modal).toBeNull();
  });

  test('settings', () => {
    expect(wrapper.vm.modal).toBeNull();
    sidebar.vm.$emit('modal:settings');
    expect(wrapper.vm.modal.name).toEqual('settings');
    expect(wrapper).toMatchSnapshot();

    wrapper.find(ModalDialogue).vm.$emit('close');
    expect(wrapper.vm.modal).toBeNull();
  });
});

test('save settings', () => {
  const wrapper = mount(HeadUp, {
    localVue,
    stubs: ['v-icon'],
  });
  const sidebar = wrapper.find(Sidebar);

  sidebar.vm.$emit('modal:settings');
  const settings = wrapper.find(SettingsScreen);

  expect(wrapper.vm.state.settings.smoothScrolling.value).toBe(true);
  settings.find('[name="smoothScrolling"]').trigger('click');
  expect(wrapper.vm.state.settings.smoothScrolling.value).toBe(false);
});

describe('persist changes', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithProps();
  });

  test('restore state from local storage', () => {
    expect(localStorage.getItem).toHaveBeenCalledWith('headUp');
  });

  test('enabled persistence', () => {
    wrapper.setData({
      state: {
        activeBoardId: '2',
      },
    });

    expect(localStorage.setItem).toHaveBeenCalledWith(
      'headUp',
      JSON.stringify(wrapper.vm.state),
    );
  });

  test('disabled persistence', () => {
    wrapper.setData({
      persistedState: JSON.stringify(wrapper.vm.state),
      state: {
        settings: {
          persistState: {
            value: false,
          },
        },
      },
    });
    wrapper.setData({
      state: {
        activeBoardId: '2',
      },
    });

    const expectedState = merge(JSON.parse(wrapper.vm.persistedState), {
      settings: {
        persistState: {
          value: false,
        },
      },
    });

    expect(localStorage.setItem).toHaveBeenCalledWith(
      'headUp',
      JSON.stringify(expectedState),
    );
  });
});

describe('respond to keystrokes', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithProps({
      attachToDocument: true,
    });
  });

  test('next board', () => {
    expect(wrapper.vm.state.activeBoardId).toEqual('1');
    wrapper.trigger('keydown', {
      key: 'j',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('2');
    wrapper.trigger('keydown', {
      key: 'j',
    });
    wrapper.trigger('keydown', {
      key: 'j',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('1');
  });

  test('previous board', () => {
    wrapper.trigger('keydown', {
      key: 'k',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('3');

    wrapper.trigger('keydown', {
      key: 'k',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('2');
  });

  test('toggle sidebar', () => {
    expect(wrapper.vm.state.showSidebar).toEqual(true);
    wrapper.trigger('keydown', {
      key: 's',
    });
    expect(wrapper.vm.state.showSidebar).toEqual(false);
  });

  test('toggle edit', () => {
    expect(wrapper.vm.state.editMode).toEqual(false);
    wrapper.trigger('keydown', {
      key: 'e',
    });
    expect(wrapper.vm.state.editMode).toEqual(true);
  });

  test('add board', () => {
    expect(wrapper.vm.state.boards.length).toEqual(3);
    wrapper.trigger('keydown', {
      key: 'a',
    });
    expect(wrapper.vm.state.boards.length).toEqual(4);
  });

  //  TODO: Keystroke combinations
});

test('focus trap', () => {
  const wrapper = mountWithSlot();

  wrapper.setData({
    state: {
      editMode: true,
    },
  });

  expect(ally.maintain.tabFocus).toHaveBeenCalled();
});
