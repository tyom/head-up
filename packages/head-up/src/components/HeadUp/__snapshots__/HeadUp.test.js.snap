// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal dialogues help 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <headupboards-stub boards="">
      <div class="Board">
        <!---->
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </headupboards-stub>
  </div>
  <modaldialogue-stub heading="Keyboard help">
    <helpscreen-stub></helpscreen-stub>
    <!---->
  </modaldialogue-stub>
</div>
`;

exports[`modal dialogues settings 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <headupboards-stub boards="">
      <div class="Board">
        <!---->
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </headupboards-stub>
  </div>
  <modaldialogue-stub heading="Settings">
    <!---->
    <settingsscreen-stub settings="[object Object]"></settingsscreen-stub>
  </modaldialogue-stub>
</div>
`;

exports[`render default 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <headupboards-stub boards=""></headupboards-stub>
  </div>
  <!---->
</div>
`;

exports[`render in edit mode 1`] = `
<div class="HeadUp _edit">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <headupboards-stub boards=""></headupboards-stub>
  </div>
  <!---->
</div>
`;

exports[`render with props 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <div class="HeadUpBoards">
      <div class="Board">
        <!---->
        <div class="cells u-grid-2-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #2</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="Board">
        <!---->
        <div class="cells u-grid-1-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #3</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="Board">
        <!---->
        <div class="cells u-grid-1-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #4</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`render with slot 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <headupboards-stub boards="">
      <div class="Board">
        <!---->
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="container">
              <div class="body"></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </headupboards-stub>
  </div>
  <!---->
</div>
`;

exports[`render without sidebar 1`] = `
<div class="HeadUp">
  <!---->
  <div class="main">
    <headupboards-stub boards=""></headupboards-stub>
  </div>
  <!---->
</div>
`;
